quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -160,671 +164,714 @@ fn mul_matches(left_ax: &Expression, right_ax: &Expression) -> bool {
+                let new = simplify_infix(&new_ab, InfixOperator::Plus, c, limit - 1);
+                min_by_key(original, new, size)
+            }
+
+            // Addition associative, left: (a + b) + c = a + (b + c), pick the shorter
+            (
+                ref left @ Expression::Infix(InfixExpression {
+                    left: ref a,
+                    operator: InfixOperator::Plus,
+                    right: ref b,
+                }),
+                InfixOperator::Plus,
+                ref c,
+            ) => {
+                let original = mul!(left.clone(), c.clone());
+                let bc = simplify_infix(b, InfixOperator::Plus, c, limit - 1);
+                let new = simplify_infix(a, InfixOperator::Plus, &bc, limit - 1);
+                min_by_key(original, new, size)
+            }
+
+            // Multiplication associative, right: a * (b * c) = (a * b) * c, pick the shorter
+            (
+                ref a,
                 InfixOperator::Star,
-                &new_multiplicand,
-                limit.saturating_sub(1),
-            );
-            min_by_key(original, new, size)
-        }
-
-        // Mul inside Div on right
-        (
-            ref numerator,
-            InfixOperator::Slash,
-            ref denominator @ Expression::Infix(InfixExpression {
-                left: ref multiplier,
-                operator: InfixOperator::Star,
-                right: ref multiplicand,
-            }),
-        ) => {
-            let original = div!(numerator.clone(), denominator.clone());
-            let new_multiplier = simplify_infix(
-                numerator,
+                ref right @ Expression::Infix(InfixExpression {
+                    left: ref b,
+                    operator: InfixOperator::Star,
+                    right: ref c,
+                }),
+            ) => {
+                let original = mul!(a.clone(), right.clone());
+                let ab = simplify_infix(a, InfixOperator::Star, b, limit - 1);
+                let new = simplify_infix(&ab, InfixOperator::Star, c, limit - 1);
+                min_by_key(original, new, size)
+            }
+
+            // Multiplication associative, left: (a * b) * c = a * (b * c), pick the shorter
+            (
+                ref left @ Expression::Infix(InfixExpression {
+                    left: ref a,
+                    operator: InfixOperator::Star,
+                    right: ref b,
+                }),
+                InfixOperator::Star,
+                ref c,
+            ) => {
+                let original = mul!(left.clone(), c.clone());
+                let bc = simplify_infix(b, InfixOperator::Star, c, limit - 1);
+                let new = simplify_infix(a, InfixOperator::Star, &bc, limit - 1);
+                min_by_key(original, new, size)
+            }
+
+            // Subtraction "associative" (not really), right: a - (b - c) = (a + c) - b
```

Comment:
Could merge with the below case

Reply:
Same comment as before re: reconstructing the original